website/static/js/nodesPrivacy.js

```diff
@@ -0,0 +1,281 @@
+        title: nodeTree.node.title,
+        canWrite: nodeTree.node.can_write,
+        changed: false
+    };
+
+    if (nodeTree.children) {
+        for (i in nodeTree.children) {
+            nodesOriginal = getNodesOriginal(nodeTree.children[i], nodesOriginal);
+        }
+    }
+    return nodesOriginal;
+}
+
+    /**
+     * patches all the nodes in a changed state
+     * uses API v2 bulk requests
+     */
+function patchNodesPrivacy(nodes) {
+    var nodesV2Url = window.contextVars.apiV2Prefix + 'nodes/';
+    var nodesPatch = nodes.splice(0,10).map(function (node) {
+        return {
+            'type': 'nodes',
+            'id': node.id,
+            'attributes': {
+                'public': node.public
+            }
+        };
+    });
+    return $.ajax({
+        url: nodesV2Url,
+        type: 'PATCH',
+        dataType: 'json',
+        contentType: 'application/vnd.api+json; ext=bulk',
+        crossOrigin: true,
+        xhrFields: {withCredentials: true},
+        processData: false,
+        data: JSON.stringify({
+            data: nodesPatch
+        })
+    }).done(function (response) {
+        if (nodes.length === 0) {
+            return;
+        }
+        return patchNodesPrivacy(nodes);
+    }).fail(function(xhr, status, error) {
+        $osf.growl('Error', 'Unable to update project privacy');
+        Raven.captureMessage('Could not PATCH project settings.', {
+            url: nodesV2Url, status: status, error: error
+        });
+    });
+}
+
+    /**
+     * view model which corresponds to nodes_privacy.mako (#nodesPrivacy)
+     *
+     * @type {NodesPrivacyViewModel}
+     */
+var NodesPrivacyViewModel = function(data, parentIsPublic) {
+    var self = this;
+    self.WARNING = 'warning';
+    self.SELECT = 'select';
+    self.CONFIRM = 'confirm';
+
+
+    var treebeardUrl = data.node.api_url  + 'get_node_tree/';
+    self.nodesOriginal = {};
+    self.loadingResults = ko.observable(false);
+    //state of current nodes
+    self.nodesState = ko.observableArray();
+    self.nodesState.subscribe(function(newValue) {
+        m.redraw(true);
+    });
+    //original node state on page load
+    self.nodesChangedPublic = ko.observableArray([]);
+    self.nodesChangedPrivate = ko.observableArray([]);
+    self.hasChildren = ko.observable(false);
+    $('#nodesPrivacy').on('hidden.bs.modal', function () {
+        self.clear();
+    });
+
+    self.loadingResults(true); // enables spinner
+    /**
+     * get node tree for treebeard from API V1
+     */
+    self.fetchNodeTree = function() {
+        return $.ajax({
+            url: treebeardUrl,
+            type: 'GET',
+            dataType: 'json'
+        }).done(function(response) {
+            var i = 0;
+            self.nodesOriginal = getNodesOriginal(response[0], self.nodesOriginal);
+            for (var key in self.nodesOriginal) {
+                i++;
+            }
+            if (i > 1) {
+                self.hasChildren(true);
```

Comment:
`hasChildren` is derived from `self.nodesOriginal`, so it should probably be a `ko.computed` rather than an `observable`.